pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -707,73 +709,96 @@ private void completeConnect(int clientProtoVersion, String clientVersion, boole
-    private State doAuthentication(AuthData clientData,
-                                   int clientProtocolVersion,
-                                   String clientVersion) throws Exception {
-
+    private CompletableFuture<Void> doAuthenticationAsync(AuthData clientData, int clientProtocolVersion,
+                                                          String clientVersion) {
         // The original auth state can only be set on subsequent auth attempts (and only
         // in presence of a proxy and if the proxy is forwarding the credentials).
         // In this case, the re-validation needs to be done against the original client
-        // credentials.
-        boolean useOriginalAuthState = (originalAuthState != null);
-        AuthenticationState authState =  useOriginalAuthState ? originalAuthState : this.authState;
-        String authRole = useOriginalAuthState ? originalPrincipal : this.authRole;
-        AuthData brokerData = authState.authenticate(clientData);
-
-        if (log.isDebugEnabled()) {
-            log.debug("Authenticate using original auth state : {}, role = {}", useOriginalAuthState, authRole);
-        }
+        // credentials, but we only can new an authentication state, because some authentication data(TLS, SASL)
+        // based on outside service.
+        // If we can get the role from the authentication sate, the global variable need to be updated.
 
-        if (authState.isComplete()) {
-            // Authentication has completed. It was either:
-            // 1. the 1st time the authentication process was done, in which case we'll send
-            //    a `CommandConnected` response
-            // 2. an authentication refresh, in which case we need to refresh authenticationData
-
-            String newAuthRole = authState.getAuthRole();
-
-            // Refresh the auth data.
-            this.authenticationData = authState.getAuthDataSource();
+        boolean useOriginalAuthState = (originalAuthState != null);
+        if (state == State.Connected) {
+            // For auth challenge, the authentication state requires to be updated.
             if (log.isDebugEnabled()) {
-                log.debug("[{}] Auth data refreshed for role={}", remoteAddress, this.authRole);
+                log.debug("Refreshing authenticate state, original auth state: {}, original auth role: {}, "
+                                + "auth role: {}",
+                        useOriginalAuthState, originalPrincipal, authRole);
             }
-
-            if (!useOriginalAuthState) {
-                this.authRole = newAuthRole;
+            try {
+                if (useOriginalAuthState) {
+                    originalAuthState =
+                            originalAuthenticationProvider.newAuthState(clientData, remoteAddress, sslSession);
+                } else {
+                    authState = authenticationProvider.newAuthState(clientData, remoteAddress, sslSession);
```

Comment:
This does not feel right based on the `AuthenticationState` interface, which provides hooks for calls to `authenticate`. Can you provide additional motivation for this change?

Reply:
I think the underlying problem is that calling `authenticate` does not update the `AuthenticationDataSource` in our token authentication provider. One solution is to recreate an auth state object. However, this has not been the implicit contract in the `ServerCnx` for as long as we've had the `AuthenticationState`. The contract has been that the `AuthenticationState` lasts the whole life of the connection. I think we should update the `TokenAuthenticationState` so that `authenticate` updates the `AuthenticationDataSource`.

Reply:
Right. We can fix built-in plugins easily, but we also need to consider third-party plugins.

Reply:
I actually don't think third-party plugins are going to be a problem here. The `AuthenticationState` object has lived the whole life of the `ServerCnx` already, so updating the `TokenAuthenticationState#authenticate` implementation shouldn't be a problem. Or, when you say "third-party plugins", are you referencing to those plugins relying on the `TokenAuthenticationState`? Even so, I think it's a pretty easy change, and I'd even consider it a bug since the `AuthenticationDataSource` is static even though we try to update it in the `ServerCnx`.

Reply:
I created https://github.com/apache/pulsar/pull/19282 to start a concrete discussion.

Reply:
Third-party plugins are the custom plug-in from users.

Reply:
> I worry that breaks the user's plugin here if we don't recreate `AuthenticationState`.

We don't recreate the auth state now. Is there a specific case that you're worried about breaking?

Reply:
> We don't recreate the auth state now. Is there a specific case that you're worried about breaking?

Sure.